### ISSUE TITLE
`nix run` invokes command with system `bash`?

### ISSUE THREAD

**Issue by sellout** [NixOS/nix#9488]
**Describe the bug**

In some convoluted situations, `nix run` can end up executing a command with the wrong interpreter.

tl;dr
- BSD-derived kernels (like darwin) don't allow a script (with a shebang, as opposed to a binary) to be used as a shebang command interpreter.
- When this happens (as I understand it) the exec dispatcher (e.g., the shell) sometimes tries to recover. Bash recovers by simply interpreting the file itself.
- Under `nix run` on MacOS, this results in Apple's bundled Bash 3.2 being used as the interpreter (even if a newer Bash is on the `PATH`).

The context for this is that I have a wrapper script around bash that I use in shebang lines. It generally works well, but there was a case I ran into that really threw me.

Basically, BSD-derived kernels (like MacOS) don't allow scripts as shebang interpreters. They fail with `ENOEXEC` and then things get weird. ZSH (the default Mac shell) will try to run the script with the correct interpreter, completely hiding the kernel issue. Bash just says "I'll be the interpreter!" If you had a wrapper around Python or something, that would quickly lead to an error, no big deal. But my wrapper was around Bash, so _most of the time_ I couldn't tell that my wrapper was being ignored.

But with `nix run` it ends up using the system Bash, which is ancient, and so the script I was running (which expected to be run by a Nix-provided Bash 5.2 wrapped in a small script) would error out. This _only_ happens with `nix run` and it uses 3.2 even if the environment it was run in has a newer Bash.

**Steps To Reproduce**

I put up a branch of a project with some extra debugging info. It prints out the version of Bash being used. This is more obvious on a non-NixOS system (like a Mac in this case), where the system’s Bash is more likely to be a different version than the shebang line, but it should be reproducible enough anywhere.

```
$ nix run github:sellout/project-manager/nix-run-repro
/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager
3.2.57(1)-release
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: conditional binary operator expected
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: syntax error near `NO_COLOR'
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: `  if [[ ! -v NO_COLOR && -t 1 ]]; then'
/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager: line 38: conditional binary operator expected
/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager: line 38: syntax error near `VERBOSE'
/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager: line 38: `  if [[ -v VERBOSE ]]; then'
```

Those errors after the Bash version are because MacOS’ included Bash (3.2) is too old to run the script. It doesn’t support `-v`. I don’t think I would have noticed otherwise.

Alternatively, running under `nix develop` still ignores the shebang, but the shell is a modern Bash, so it runs successfully:
```
$ nix develop github:sellout/project-manager/nix-run-repro

❄️$ project-manager
/nix/store/x1xxsh1gp6y389hyl40a0i74dkxiprl7-bash-5.2-p15/bin/bash
5.2.15(1)-release
…
```

**Expected behavior**

I expect it to either error, reporting the `ENOEXEC` failure, or to behave the same as under ZSH (or on Linux):
```
$ nix path-info github:sellout/project-manager/nix-run-repro
/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager

$ /nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager
running strict-bash
strict-bash version: 5.2.15(1)-release
/nix/store/xhvxpfkb3xmicav5wdx9xfyiaw452mbl-bash-interactive-5.2-p15/bin/bash
5.2.15(1)-release
…
```
Running the script directly includes the debug output (`running strict-bash` `strict-bash version: 5.2.15(1)-release`) I added to the shebang interpreter to illustrate this issue. You can see how the `nix run` and `nix develop` versions don’t include that, because it never runs that interpreter. The last two lines are the debugging output I added to `project-manager` that shows it’s now using the version of Bash wrapped in the interpreter script, and not the old MacOS version.

I would also accept it just using a Nix-provided Bash. That just makes the incorrect behavior more invisible in my case, but at least it's a bit more consistent.

**`nix-env --version` output**
```
nix (Nix) 2.20.0pre20231128_a8fea5a
```

**Additional context**

This was hard for me to track down. I couldn't figure out why _some_ uses of the script worked while others didn’t, despite my efforts to make it reproducible. I expected `nix run` to be the _most_ hermetic way of running this (and in a way, it is – it uses the system Bash despite my best efforts), at least using a Nix-provided Bash.

Here is an issue on my own project: sellout/bash-strict-mode#10

and where I found the most helpful context: https://stackoverflow.com/a/9988275/5090651

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by abathur** [comment]
I'm not sure you'll find a satisfying answer from the nix side, here.

IIRC, when macOS encounters an invalid shebang, it tries to fall back on a default interpreter. I'm not sure if there's a good succinct reference to the behavior, but I imagine it's searching PATH and then falling back on either a hardcoded locations or a hardcoded fallback PATH.

Consider:
```
$ result/bin/project-manager --help
/run/current-system/sw/bin/bash
5.2.15(1)-release
Usage: result/bin/project-manager [OPTION] COMMAND
...

$ /usr/bin/env -i result/bin/project-manager --help
/bin/sh
3.2.57(1)-release
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: conditional binary operator expected
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: syntax error near `NO_COLOR'
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: `  if [[ ! -v NO_COLOR && -t 1 ]]; then'
result/bin/project-manager: line 38: conditional binary operator expected
result/bin/project-manager: line 38: syntax error near `VERBOSE'
result/bin/project-manager: line 38: `  if [[ -v VERBOSE ]]; then'
```

If that thesis is right, I imagine there isn't much Nix can do here aside from modify the PATH (which is just as likely to cause its own hermeticity problems). I'd personally try to avoid behavior that'll trigger the fallback.

As an aside, you might find a project I develop, [resholve](https://github.com/abathur/resholve), of some utility for working with Shell in the Nix ecosystem. Here's a description of the Nix API: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/misc/resholve/README.md

**Comment by sellout** [comment]
> ```
> $ /usr/bin/env -i result/bin/project-manager --help
> /bin/sh
> 3.2.57(1)-release
> ```

I think I see – you’re saying that `nix run` is basically using an empty env to be as consistent as possible already, and trying to "fix" this by adding some Bash to the PATH is probably the wrong thing.

> I'd personally try to avoid behavior that'll trigger the fallback.

I agree, which is why my preferred solution would be to report the `ENOEXEC`, rather than the system trying to guess what interpreter to use (at least ZSH does it successfully, but I’d still be happier if it just told me I was doing something wrong – and ideally what). It was quite baffling to get this inconsistent behavior.

Looking at https://github.com/NixOS/nix/blob/02bd821f2e71372d31bbe6700bd68086cc2ee70a/src/nix/run.cc#L275
It seems like [`execvp` and friends handle `ENOEXEC` by trying `/bin/sh`](https://www.man7.org/linux/man-pages/man3/execvp.3.html#DESCRIPTION). However, since (I think) `cmd` is always an absolute path to the Nix store, we don’t need the `PATH` searching that `execvp` does, so perhaps that line could become
```c++
execv(cmd.c_str(), stringsToCharPtrs(args).data());
```
avoiding the special error handling used for the `p` functions and resulting in the failure I'd like to see. I’ll hopefully get a chance to try this out tomorrow.

> As an aside, you might find a project I develop, [resholve](https://github.com/abathur/resholve), of some utility for working with Shell in the Nix ecosystem. Here's a description of the Nix API: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/misc/resholve/README.md

Ah, I've come across resholve before, but I think too early in my Nix journey to understand how it could help me. Thanks for bringing it up!

**Comment by sellout** [comment]
Ok, that seemed to do it.
```
$ nix run github:sellout/project-manager/nix-run-repro
/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager
3.2.57(1)-release
path '/Users/greg/Projects/community/bash-strict-mode' does not contain a 'flake.nix', searching up
error: could not find a flake.nix file
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: conditional binary operator expected
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: syntax error near `NO_COLOR'
/nix/store/h38nb91yhksxc6vqivs8nqcs1m9kqfhm-project-manager.bash: line 19: `  if [[ ! -v NO_COLOR && -t 1 ]]; then'
/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager: line 38: conditional binary operator expected
/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager: line 38: syntax error near `VERBOSE'

$ nix shell github:sellout/nix/nix-run-execv

$ nix run github:sellout/project-manager/nix-run-repro
error: unable to execute '/nix/store/bik3y0in6jghz2i8a81r8ncpz1i5bbac-project-manager/bin/project-manager': Exec format error
```
